internal/component/database_observability/mysql/collector/query_sample.go

```diff
@@ -0,0 +1,290 @@
+	instanceKey     string
+	collectInterval time.Duration
+	entryHandler    loki.EntryHandler
+	sqlParser       parser.Parser
+
+	logger  log.Logger
+	running *atomic.Bool
+	ctx     context.Context
+	cancel  context.CancelFunc
+
+	lastSampleSeenTimestamp float64
+}
+
+func NewQuerySample(args QuerySampleArguments) (*QuerySample, error) {
+	c := &QuerySample{
+		dbConnection:    args.DB,
+		instanceKey:     args.InstanceKey,
+		collectInterval: args.CollectInterval,
+		entryHandler:    args.EntryHandler,
+		logger:          log.With(args.Logger, "collector", QuerySampleName),
+		running:         &atomic.Bool{},
+	}
+
+	if args.UseTiDBParser {
+		c.sqlParser = parser.NewTiDBSqlParser()
+	} else {
+		c.sqlParser = parser.NewXwbSqlParser()
+	}
+
+	return c, nil
+}
+
+func (c *QuerySample) Name() string {
+	return QuerySampleName
+}
+
+func (c *QuerySample) Start(ctx context.Context) error {
+	level.Debug(c.logger).Log("msg", QuerySampleName+" collector started")
+
+	c.running.Store(true)
+	ctx, cancel := context.WithCancel(ctx)
+	c.ctx = ctx
+	c.cancel = cancel
+
+	if err := c.setLastSampleSeenTimestamp(c.ctx); err != nil {
+		level.Error(c.logger).Log("msg", "failed to set last sample seen timestamp", "err", err)
+		return err
+	}
+
+	go func() {
+		defer func() {
+			c.Stop()
+			c.running.Store(false)
+		}()
+
+		ticker := time.NewTicker(c.collectInterval)
+
+		for {
+			if err := c.fetchQuerySamples(c.ctx); err != nil {
+				level.Error(c.logger).Log("msg", "collector error", "err", err)
+			}
+
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-ticker.C:
+				// continue loop
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (c *QuerySample) Stopped() bool {
+	return !c.running.Load()
+}
+
+// Stop should be kept idempotent
+func (c *QuerySample) Stop() {
+	c.cancel()
+}
+
+// setLastSampleSeenTimestamp queries the database for the latest sample seen timestamp so that upon startup we don't collect
+// samples that may have been collected previously.
+func (c *QuerySample) setLastSampleSeenTimestamp(ctx context.Context) error {
+	rs, err := c.dbConnection.QueryContext(ctx, selectLatestTimerEnd)
+	if err != nil {
+		return fmt.Errorf("failed to fetch last sample seen timestamp: %w", err)
+	}
+	defer rs.Close()
+
+	for rs.Next() {
+		if err := rs.Err(); err != nil {
+			return fmt.Errorf("failed to iterate rows: %w", err)
+		}
+		var ts sql.NullFloat64
+		err := rs.Scan(&ts)
+		if err != nil {
+			return fmt.Errorf("failed to scan row: %w", err)
+		}
+		if !ts.Valid {
+			return fmt.Errorf("no valid timestamp found: %w", err)
+		}
+		c.lastSampleSeenTimestamp = ts.Float64
+	}
+	return nil
+}
+
+func (c *QuerySample) fetchQuerySamples(ctx context.Context) error {
+	rs, err := c.dbConnection.QueryContext(ctx, selectQuerySamples, c.lastSampleSeenTimestamp)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to fetch history table samples", "err", err)
+		return err
+	}
+	defer rs.Close()
+
+	for rs.Next() {
+		row := struct {
```

Comment:
Nit: why don't we define a named struct?

Reply:
Because it's not used anywhere else, I thought we could keep it within the function to keep the (private) package surface smaller.